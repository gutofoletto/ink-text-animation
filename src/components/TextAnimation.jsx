import React from 'react';
import chalkAnimation from 'chalk-animation';
import PropTypes from 'prop-types';

const delays = {
  rainbow: 15,
  pulse: 16,
  glitch: 55,
  radar: 50,
  neon: 500,
};

class TextAnimation extends React.Component {
  constructor(props, context) {
    super(props, context);

    this.timeout = null;
    this.animation = chalkAnimation[props.name]('').stop();
    this.state = {
      frame: '',
    };
  }

  componentDidMount() {
    this.start();
  }

  componentWillReceiveProps(nextProps) {
    const { name } = this.props;

    if (name !== nextProps.name) {
      this.animation = chalkAnimation[nextProps.name]('');
      this.animation.stop();
    }
  }

  componentWillUnmount() {
    this.stop();
  }

  start() {
    const { name, speed } = this.props;

    this.update();

    // Using `setTimeout` instead of `setInterval`
    // since the animation can change, thus its delay.
    this.timeout = setTimeout(() => {
      this.start();
    }, delays[name] / speed);
  }

  update() {
    const { children } = this.props;
    const frame = this.animation
      .replace(children)
      .frame()
      .replace(/^\u001B\[(\d)F\u001B\[G\u001B\[2K/, '');

    this.setState({ frame });
  }

  stop() {
    clearTimeout(this.timeout);
    this.timeout = null;
    this.animation.destroy();
  }

  render() {
    return this.state.frame;
  }
}

TextAnimation.defaultProps = {
  name: 'rainbow',
  speed: 1,
};

TextAnimation.propTypes = {
  children: PropTypes.oneOfType([PropTypes.string]).isRequired,
  name: PropTypes.oneOf(['rainbow', 'pulse', 'glitch', 'radar', 'neon']),
  speed: PropTypes.number,
};

export default TextAnimation;
