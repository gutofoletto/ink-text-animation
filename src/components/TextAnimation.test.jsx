import React from 'react';
import { render } from 'ink-testing-library';
import chalk from 'chalk';
import TextAnimation from './TextAnimation';

chalk.enabled = true;
chalk.level = 2;

const text = 'Lorem ipsum';

describe('<TextAnimation />', () => {
  it('should render with default animation', () => {
    const { lastFrame, unmount } = render(<TextAnimation>{text}</TextAnimation>);
    expect(lastFrame()).toMatchSnapshot();
    unmount();
  });

  it('should render with rainbow animation', () => {
    const { lastFrame, unmount } = render(<TextAnimation name="rainbow">{text}</TextAnimation>);
    expect(lastFrame()).toMatchSnapshot();
    unmount();
  });

  it('should render with pulse animation', () => {
    const { lastFrame, unmount } = render(<TextAnimation name="pulse">{text}</TextAnimation>);
    expect(lastFrame()).toMatchSnapshot();
    unmount();
  });

  it('should render with radar animation', () => {
    const { lastFrame, unmount } = render(<TextAnimation name="radar">{text}</TextAnimation>);
    expect(lastFrame()).toMatchSnapshot();
    unmount();
  });

  it('should render with neon animation', () => {
    const { lastFrame, unmount } = render(<TextAnimation name="neon">{text}</TextAnimation>);
    expect(lastFrame()).toMatchSnapshot();
    unmount();
  });

  it('should render with karaoke animation', () => {
    const { lastFrame, unmount } = render(<TextAnimation name="karaoke">{text}</TextAnimation>);
    expect(lastFrame()).toMatchSnapshot();
    unmount();
  });

  describe('when using the glitch animation', () => {
    it('should render some random characters', () => {
      const { lastFrame, unmount } = render(<TextAnimation name="glitch">{text}</TextAnimation>);
      expect([...lastFrame()].some(char => text.includes(char))).toBeTruthy();
      unmount();
    });
  });

  describe('when changing the text prop', () => {
    it('should render new text', () => {
      const { lastFrame, unmount, rerender } = render(<TextAnimation>{text}</TextAnimation>);
      expect(lastFrame()).toMatchSnapshot();
      rerender(<TextAnimation>New Text</TextAnimation>);
      expect(lastFrame()).toMatchSnapshot();
      unmount();
    });
  });
});
