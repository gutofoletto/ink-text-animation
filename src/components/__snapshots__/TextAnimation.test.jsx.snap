// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextAnimation /> should render with default animation 1`] = `"[38;2;255;21;0mL[39m[38;2;255;0;148mo[39m[38;2;192;0;255mr[39m[38;2;23;0;255me[39m[38;2;0;147;255mm[39m [38;2;0;255;194mi[39m[38;2;0;255;24mp[39m[38;2;145;255;0ms[39m[38;2;255;195;0mu[39m[38;2;255;25;0mm[39m"`;

exports[`<TextAnimation /> should render with karaoke animation 1`] = `"[37mLorem ipsum[39m"`;

exports[`<TextAnimation /> should render with neon animation 1`] = `"[1m[38;2;213;70;242mLorem ipsum[39m[22m"`;

exports[`<TextAnimation /> should render with pulse animation 1`] = `"[38;2;230;230;230mL[39m[38;2;230;230;230mo[39m[38;2;230;230;230mr[39m[38;2;243;123;123me[39m[38;2;255;16;16mm[39m [38;2;247;87;87mi[39m[38;2;238;159;159mp[39m[38;2;230;230;230ms[39m[38;2;230;230;230mu[39m[38;2;230;230;230mm[39m"`;

exports[`<TextAnimation /> should render with radar animation 1`] = `"[38;2;127;127;127mL[39m"`;

exports[`<TextAnimation /> should render with rainbow animation 1`] = `"[38;2;255;21;0mL[39m[38;2;255;0;148mo[39m[38;2;192;0;255mr[39m[38;2;23;0;255me[39m[38;2;0;147;255mm[39m [38;2;0;255;194mi[39m[38;2;0;255;24mp[39m[38;2;145;255;0ms[39m[38;2;255;195;0mu[39m[38;2;255;25;0mm[39m"`;
